packages/modal-ui-js/src/lib/render-modal.ts

```diff
@@ -0,0 +1,187 @@
+import { renderNoLedgerAccountsFound } from "./components/NoLedgerAccountsFound";
+import { renderWalletConnecting } from "./components/WalletConnecting";
+import { renderWalletConnectionFailed } from "./components/WalletConnectionFailed";
+import { renderWalletNotInstalled } from "./components/WalletNotInstalled";
+import { modalState } from "./modal";
+
+export type HardwareWalletAccountState = HardwareWalletAccount & {
+  selected: boolean;
+};
+
+const getAccountIds = async (publicKey: string): Promise<Array<string>> => {
+  if (!modalState) {
+    return [];
+  }
+
+  const response = await fetch(
+    `${modalState.selector.options.network.indexerUrl}/publicKey/ed25519:${publicKey}/accounts`
+  );
+
+  if (!response.ok) {
+    throw new Error("Failed to get account id from public key");
+  }
+
+  const accountIds = await response.json();
+
+  if (!Array.isArray(accountIds) || !accountIds.length) {
+    return [];
+  }
+
+  return accountIds;
+};
+
+export const resolveAccounts = async (
+  wallet: Wallet
+): Promise<Array<HardwareWalletAccountState> | null> => {
+  if (!modalState) {
+    return [];
+  }
+
+  const publicKey = await (wallet as HardwareWallet).getPublicKey(
+    modalState.derivationPath
+  );
+  try {
+    const accountIds = await getAccountIds(publicKey);
+
+    return accountIds.map((accountId, index) => {
+      return {
+        derivationPath: modalState!.derivationPath,
+        publicKey,
+        accountId,
+        selected: index === 0,
+      };
+    });
+  } catch (e) {
+    return null;
+  }
+};
+
+export async function connectToWallet(module: ModuleState<Wallet>) {
+  if (!modalState) {
+    return;
+  }
+
+  try {
+    if (module.type === "injected" && !module.metadata.available) {
+      return renderWalletNotInstalled(module);
+    }
+
+    if (module.metadata.deprecated) {
+      return renderWalletConnectionFailed(
+        module,
+        new Error("Wallet is depredacted")
+      );
+    }
+
+    const wallet = await module.wallet();
+
+    await renderWalletConnecting(module);
+
+    if (wallet.type === "hardware") {
+      const accounts = await resolveAccounts(wallet);
+
+      if (!accounts || accounts.length < 1) {
+        return renderNoLedgerAccountsFound(module);
+      }
+
+      if (accounts.length === 1) {
+        return renderLedgerAccountsOverviewList(module, accounts, accounts);
+      } else {
+        return renderLedgerSelectAccount(module, accounts);
+      }
+    }
+
+    await wallet.signIn({
+      contractId: modalState.options.contractId,
+      methodNames: modalState.options.methodNames,
+    });
+
+    modalState.container.children[0].classList.remove("open");
+  } catch (err) {
+    await renderWalletConnectionFailed(module, err as Error);
+  }
+}
+
+export function renderModal() {
+  if (!modalState) {
+    return;
+  }
+
+  modalState.container.innerHTML = `
+    <div class="nws-modal-wrapper">
+      <div class="nws-modal-overlay"></div>
```

Comment:
Would it be possible to close the modal when we click on the overlay (outside the modal?

Also, it would be nice to close the modal when we press `ESC` like we have in the react modal.